packages/opentelemetry-resources/src/Resource.ts

```diff
@@ -54,8 +58,50 @@ export class Resource {
      * information about the entity as numbers, strings or booleans
      * TODO: Consider to add check/validation on attributes.
      */
-    readonly attributes: ResourceAttributes
-  ) {}
+    attributes: ResourceAttributes,
+    asyncAttributesPromise?: Promise<ResourceAttributes>
+  ) {
+    this._attributes = attributes;
+    this._asyncAttributesHaveResolved = asyncAttributesPromise === undefined;
+    this._asyncAttributesPromise = asyncAttributesPromise?.then(
+      asyncAttributes => {
+        this._attributes = Object.assign({}, this._attributes, asyncAttributes);
+        this._asyncAttributesHaveResolved = true;
+        return asyncAttributes;
+      }
+    );
+    this._asyncAttributesPromise?.catch(err => {
+      diag.debug("The resource's async promise rejected: %s", err);
+      this._asyncAttributesHaveResolved = true;
+      return {};
+    });
+  }
+
+  get attributes(): ResourceAttributes {
+    return this._attributes;
+  }
+
+  /**
+   * Check if async attributes have resolved. This is useful to avoid awaiting
+   * waitForAsyncAttributes (which will introduce asynchronous behavior) when not necessary.
+   *
+   * @returns true if no async attributes promise was provided or if the promise has resolved
+   * and been merged together with the sync attributes.
+   */
+  asyncAttributesHaveResolved(): boolean {
+    return this._asyncAttributesHaveResolved;
+  }
+
+  /**
+   * Returns a promise that resolves when all async attributes have finished being added to
+   * this Resource's attributes. This is useful in exporters to block until resource detection
+   * has finished.
+   */
+  async waitForAsyncAttributes(): Promise<void> {
```

Comment:
is adding those properties spec compliant ?  also i'm not sure if we want to expose an API like this on the resource class

Reply:
This would be disallowed by https://github.com/open-telemetry/opentelemetry-specification/pull/2968

Reply:
No I think it's probably ok thinking about it more. I'll just resolve this.

Reply:
Could we, instead of exposing public methods for this, using the context to propagate the state of resource attributes resolving ? That would essentially "hide" the mechanism and make it that only our sdk have this behavior ?

Reply:
@vmarchaud is that ok with you?